src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector128_1.cs

```diff
@@ -692,6 +692,12 @@ private string ToString([StringSyntax(StringSyntaxAttribute.NumericFormat)] stri
         // New Surface Area
         //
 
+        static bool ISimdVector<Vector128<T>, T>.AnyMatches(Vector128<T> vector)
+        {
+            uint mask = vector.ExtractMostSignificantBits();
+            return (mask != 0);
```

Comment:
This one can just be `vector != Vector128<T>.Zero` right, which avoids needing to "extract" the bits at all (it can optimize down to a `ptest` or other instruction instead).

Reply:
I'm not sure I follow.
 This implementation on `AnyMatches` checks if the MSB for any of the values in the vector is set. 
`vector != Vector128<T>.Zero` will check if any of the values themselves are set(not just the Most Significant).

Do I understand the API incorrectly?

Reply:
There is a functional difference between the two yes and it ultimately comes down to how we want to think about "matches" in the cross platform sense.

For x86/x64 in particular, it's historically been "most significant bit set" because that's what `movmsk` extracts. However, the actual product of comparison APIs and the like is that you get either `0` or `AllBitsSet` on a per-element basis. For other platforms, like Arm64, there is just the `0` or `AllBitsSet` on a per-element basis. Avx512 and SVE where you get the built-in masking registers have changed things a bit, as now you just get 1 bit per element and converting that back to a vector gives you `0` or `AllBitsSet` on a per-element basis.

The fast and most xplat compatible thing here is then to functionally treat "match" vs "no match" much like you would `bool` where `0` is `false` (no match) and `non 0` is `true` (match). This works very well for all except a small subset of specialized x86/x64 only scenarios that truly need to check the most significant bit.

So, by doing this as `vector != Vector128<T>.Zero` we can optimize the code down to an efficient `ptest` on x86/x64 and an equivalent on other platforms like Arm64. If a user truly needed to only check the most significant bit, they could do something like `(vector & Vector128.Create(0x8000_0000)) != Vector128.Zero` and we could then add the appropriate pattern matching to optimize that down to `movmsk` and check on x86/x64.

Reply:
That would mean the required 'HasMatch' implementation and what I need for encoding are 2 different things.

 What I am looking to do is determine the following :  _**A byte that is in the range from 0 to 127 is valid, and other values are invalid**_

That would mean 
1. Switching `HasMatch ` to what you pointed out
2. Change following implementation to do the `(vector & Vector128.Create(0x8000_0000)) != Vector128.Zero` version : https://github.com/dotnet/runtime/pull/99982/files#diff-6b4906abc01dc4699f348f7c1df72e2f640f240aa31ea67cd47642221b2021f5R2200-R2208

Reply:
I think that should be fine.

`movmsk` takes 3-5 cycles (9 on Alder Lake-E cores), depending on if its `XMM` or `YMM`. That has to be followed up by a `cmp` and a `branch`.

`ptest` takes 4-6 cycles (9-10 on Alder Lake-E cores) and so should effectively be the same cost (as `movmsk; cmp`), but with 1 fewer instruction.

For AVX512 and latest, you have to do `ptestmb` (3 cycles) followed by `ktest` (or `kortest`). But that is typically lower latency, even though its 2 instructions.

Reply:
Made the change. 

Checked both locally on a Cascade Lake  - it's what you predicted

**1.** 
```
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static bool doStuffVec(Vector128<byte> vector)
{
    return ((vector & Vector128.Create((byte)0b1000_0000)) != Vector128<byte>.Zero);
}
```

is

```
G_M000_IG01:                ;; offset=0x0000

G_M000_IG02:                ;; offset=0x0000
       vmovups  xmm0, xmmword ptr [rcx]
       vptest   xmm0, xmmword ptr [reloc @RWD00]
       setne    al
       movzx    rax, al

G_M000_IG03:                ;; offset=0x0013
       ret

RWD00   dq      8080808080808080h, 8080808080808080h
```

**2.**
```
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static bool doStuffVec(Vector128<byte> vector)
{
    return (vector.ExtractMostSignificantBits() != 0);
}
```

is

```
G_M000_IG01:                ;; offset=0x0000

G_M000_IG02:                ;; offset=0x0000
       vmovups  xmm0, xmmword ptr [rcx]
       vpmovmskb eax, xmm0
       test     eax, eax
       setne    al
       movzx    rax, al

G_M000_IG03:                ;; offset=0x0010
       ret
```

Reply:
There is a bug in there. I'll fix it and get CI running again

Reply:
The bug is fixed but I want to run perf again to see if the numbers remain the same. ICX don't show much improvement. Will tests on SPR or later to get more data point

Reply:
@tannergooding Can you please review it now